### PR TITLE
Fix flaky isolation2 test.

### PR DESCRIPTION
**Pull request by higuoxing**
Currently, isolation2/test_rejectmap.sql is flaky if we run isolation2 test multiple times. That's because we set the GUC 'diskquota.hard_limit' to 'on' in test_postmaster_restart.sql and forget to set it to 'off'. In the next following runs, the hard limit is enabled and the QD will continuously dispatch reject map to segment servers. However, test_rejectmap.sql requires the hard limit being disabled because we're dispatching rejectmap by ourselves or the dispatched rejectmap will be cleared by QD.

This patch adds a new injection point to prevent QD from dispatching rejectmap to make test_rejectmap.sql stateless. This patch also set 'diskquota.hard_limit' to 'off' when test_postmaster_restart.sql finishes.

Failed test:

```diff
 -- Session 1 will return and emit an error message saying that the quota limit is exceeded on seg0.
 1<:  <... completed>
-[hardlimit] schema's disk space quota exceeded
+INSERT 10000
 
 -- Clean up the rejectmap on seg0.
 SELECT diskquota.refresh_rejectmap( ARRAY[]::diskquota.rejectmap_entry[], ARRAY[]::oid[]) FROM gp_dist_random('gp_id') WHERE gp_segment_id=0;
@@ -192,7 +197,7 @@
 
 -- Session 1 will return and emit an error message saying that the quota limit is exceeded on seg0.
 1<:  <... completed>
-[hardlimit] schema's disk space quota exceeded
+INSERT 10000
 
 -- Clean up the rejectmap on seg0.
 SELECT diskquota.refresh_rejectmap( ARRAY[]::diskquota.rejectmap_entry[], ARRAY[]::oid[]) FROM gp_dist_random('gp_id') WHERE gp_segment_id=0;
@@ -231,7 +236,7 @@
  Success:                 
 (1 row)
 1<:  <... completed>
-[hardlimit] tablespace-schema's disk space quota exceeded
+INSERT 10000
 -- Clean up the rejectmap on seg0.
 SELECT diskquota.refresh_rejectmap( ARRAY[]::diskquota.rejectmap_entry[], ARRAY[]::oid[]) FROM gp_dist_random('gp_id') WHERE gp_segment_id=0;
  refresh_rejectmap 
@@ -270,7 +275,7 @@
  Success:                 
 (1 row)
 1<:  <... completed>
-[hardlimit] tablespace-role's disk space quota exceeded
+INSERT 10000
 -- Clean up the rejectmap on seg0.
 SELECT diskquota.refresh_rejectmap( ARRAY[]::diskquota.rejectmap_entry[], ARRAY[]::oid[]) FROM gp_dist_random('gp_id') WHERE gp_segment_id=0;
  refresh_rejectmap 
@@ -344,17 +349,16 @@
 (1 row)
 -- Show that blocked_t7 is blocked on seg0.
 2: SELECT rel.segid, rel.relnamespace, rel.reltablespace, rel.relowner, replace_oid_with_relname(rel.relname, '/tmp/test_rejectmap.csv'::text), be.target_type, be.target_oid FROM gp_dist_random('diskquota.rejectmap') AS be, read_relation_cache_from_file('/tmp/test_rejectmap.csv') AS rel WHERE be.segid=rel.segid AND be.relnode=rel.relfilenode AND rel.relfilenode<>0;
- segid | relnamespace | reltablespace | relowner | replace_oid_with_relname | target_type     | target_oid 
--------+--------------+---------------+----------+--------------------------+-----------------+------------
- 0     | 2200         | 0             | 10       | blocked_t7               | NAMESPACE_QUOTA | 2200       
-(1 row)
+ segid | relnamespace | reltablespace | relowner | replace_oid_with_relname | target_type | target_oid 
+-------+--------------+---------------+----------+--------------------------+-------------+------------
+(0 rows)
 SELECT gp_inject_fault_infinite('check_rejectmap_by_relfilenode', 'reset', dbid) FROM gp_segment_configuration WHERE role='p' AND content=0;
  gp_inject_fault_infinite 
 --------------------------
  Success:                 
 (1 row)
 1<:  <... completed>
-[hardlimit] schema's disk space quota exceeded
+INSERT 10000
 1: ABORT;
 ABORT
 -- Clean up the rejectmap on seg0.
@@ -391,15 +395,14 @@
 2: SELECT rel.segid, rel.relnamespace, rel.reltablespace, rel.relowner, replace_oid_with_relname(rel.relname, '/tmp/test_rejectmap.csv'::text), be.target_type, be.target_oid FROM gp_dist_random('diskquota.rejectmap') AS be, read_relation_cache_from_file('/tmp/test_rejectmap.csv') AS rel WHERE be.segid=rel.segid AND be.relnode=rel.relfilenode AND rel.relfilenode<>0;
  segid | relnamespace | reltablespace | relowner | replace_oid_with_relname | target_type | target_oid 
 -------+--------------+---------------+----------+--------------------------+-------------+------------
- 0     | 2200         | 0             | 10       | blocked_t7               | ROLE_QUOTA  | 10         
-(1 row)
+(0 rows)
 SELECT gp_inject_fault_infinite('check_rejectmap_by_relfilenode', 'reset', dbid) FROM gp_segment_configuration WHERE role='p' AND content=0;
  gp_inject_fault_infinite 
 --------------------------
  Success:                 
 (1 row)
 1<:  <... completed>
-[hardlimit] role's disk space quota exceeded
+INSERT 10000
 1: ABORT;
 ABORT
 -- Clean up the rejectmap on seg0.
@@ -434,17 +437,16 @@
 (1 row)
 -- Show that blocked_t7 is blocked on seg0.
 2: SELECT rel.segid, rel.relnamespace, rel.reltablespace, rel.relowner, replace_oid_with_relname(rel.relname, '/tmp/test_rejectmap.csv'::text), be.target_type, be.target_oid FROM gp_dist_random('diskquota.rejectmap') AS be, read_relation_cache_from_file('/tmp/test_rejectmap.csv') AS rel WHERE be.segid=rel.segid AND be.relnode=rel.relfilenode AND rel.relfilenode<>0;
- segid | relnamespace | reltablespace | relowner | replace_oid_with_relname | target_type                | target_oid 
--------+--------------+---------------+----------+--------------------------+----------------------------+------------
- 0     | 2200         | 0             | 10       | blocked_t7               | NAMESPACE_TABLESPACE_QUOTA | 2200       
-(1 row)
+ segid | relnamespace | reltablespace | relowner | replace_oid_with_relname | target_type | target_oid 
+-------+--------------+---------------+----------+--------------------------+-------------+------------
+(0 rows)
 SELECT gp_inject_fault_infinite('check_rejectmap_by_relfilenode', 'reset', dbid) FROM gp_segment_configuration WHERE role='p' AND content=0;
  gp_inject_fault_infinite 
 --------------------------
  Success:                 
 (1 row)
 1<:  <... completed>
-[hardlimit] tablespace-schema's disk space quota exceeded
+INSERT 10000
 1: ABORT;
 ABORT
 -- Clean up the rejectmap on seg0.
@@ -479,17 +481,16 @@
 (1 row)
 -- Show that blocked_t7 is blocked on seg0.
 2: SELECT rel.segid, rel.relnamespace, rel.reltablespace, rel.relowner, replace_oid_with_relname(rel.relname, '/tmp/test_rejectmap.csv'::text), be.target_type, be.target_oid FROM gp_dist_random('diskquota.rejectmap') AS be, read_relation_cache_from_file('/tmp/test_rejectmap.csv') AS rel WHERE be.segid=rel.segid AND be.relnode=rel.relfilenode AND rel.relfilenode<>0;
- segid | relnamespace | reltablespace | relowner | replace_oid_with_relname | target_type           | target_oid 
--------+--------------+---------------+----------+--------------------------+-----------------------+------------
- 0     | 2200         | 0             | 10       | blocked_t7               | ROLE_TABLESPACE_QUOTA | 10         
-(1 row)
+ segid | relnamespace | reltablespace | relowner | replace_oid_with_relname | target_type | target_oid 
+-------+--------------+---------------+----------+--------------------------+-------------+------------
+(0 rows)
 SELECT gp_inject_fault_infinite('check_rejectmap_by_relfilenode', 'reset', dbid) FROM gp_segment_configuration WHERE role='p' AND content=0;
  gp_inject_fault_infinite 
 --------------------------
  Success:                 
 (1 row)
 1<:  <... completed>
-[hardlimit] tablespace-role's disk space quota exceeded
+INSERT 10000
 1: ABORT;
 ABORT
 -- Clean up the rejectmap on seg0.
@@ -534,7 +535,7 @@
  Success:                 
 (1 row)
 1<:  <... completed>
-[hardlimit] tablespace-schema's disk space quota exceeded
+INSERT 10000
 1: ABORT;
 ABORT
 -- Clean up the rejectmap on seg0.
@@ -579,7 +580,7 @@
  Success:                 
 (1 row)
 1<:  <... completed>
-[hardlimit] tablespace-role's disk space quota exceeded
+INSERT 10000
 1: ABORT;
 ABORT
 -- Clean up the rejectmap on seg0.
@@ -626,7 +627,7 @@
  Success:                 
 (1 row)
 1<:  <... completed>
-[hardlimit] schema's disk space quota exceeded
+INSERT 10000
 1: ABORT;
 ABORT
 -- Clean up the rejectmap on seg0.
@@ -674,7 +675,7 @@
  Success:                 
 (1 row)
 1<:  <... completed>
-[hardlimit] schema's disk space quota exceeded
+INSERT 10000
 1: ABORT;
 ABORT
 -- Clean up the rejectmap on seg0.
@@ -722,7 +723,7 @@
  Success:                 
 (1 row)
 1<:  <... completed>
-[hardlimit] schema's disk space quota exceeded
+INSERT 10000
 1: ABORT;
 ABORT
 -- Clean up the rejectmap on seg0.
@@ -732,9 +733,14 @@
                    
 (1 row)
```

### CODE REVIEW

File: quotamodel.c [hunk 1]
```diff
@@ -810,10 +810,15 @@ refresh_disk_quota_usage(bool is_init)
 		/* flush local table_size_map to user table table_size */
 		flush_to_table_size();
 		/* copy local reject map back to shared reject map */
-		bool reject_map_changed = flush_local_reject_map();
+		bool reject_map_changed        = flush_local_reject_map();
+		bool should_dispatch_rejectmap = diskquota_hardlimit && (reject_map_changed || hasActiveTable);
+#ifdef FAULT_INJECTOR
+		if (SIMPLE_FAULT_INJECTOR("disable_dispatch_reject_map") == FaultInjectorTypeSkip)
```
Comment (beeender) [inline quotamodel.c:816]: Is this fault injector really needed? I thought the flaky is ONLY because of the hardlimit setting.

**Review by beeender — COMMENTED** [review]
And conflict needs to be fixed.

Thanks for the fix!

File: quotamodel.c (hunk 1 above)
Comment (higuoxing) [reply quotamodel.c:816]: Yes, we can remove the faultinjector. But we will need to keep it in mind that this test should be executed with hardlimit being disabled.

**Review by beeender — APPROVED** [review]
CI needs to be retrigerred.

**Pull request merged**